### PR TITLE
Bump to add 3.12

### PR DESCRIPTION
**Pull request by Chris-Sigopt**
Add support for py3.12.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -45,6 +45,11 @@
 hyperopt_install_requires = ["hyperopt>=0.2.7"]
 lite_install_requires = ["sigoptlite>=0.1.1"]
 
+if sys.version_info[0] == 3 and sys.version_info[1] < 12:
+  pytest_requires = ["pytest==7.2.1"]
+else:
+  pytest_requires = ["pytest>=7.2.1"]
+
```
Comment (tjs-intel) [inline setup.py:52]: Newer pytests don't support older pythons?

File: test/runs/test_set_project.py [hunk 2]
```diff
@@ -16,7 +18,7 @@ def test_set_project_with_global_run():
 
 def test_set_project_without_run():
   with mock.patch("sigopt.get_run_id", mock.Mock(return_value=None)):
-    with pytest.warns(None) as warnings:
+    with warnings.catch_warnings():
+      warnings.simplefilter("error")
       set_project("test-123")
-  assert len(warnings) == 0
```
Comment (tjs-intel) [inline test/runs/test_set_project.py:21]: Are these equivalent?

Comment (Chris-Sigopt) [reply test/runs/test_set_project.py:21]: https://stackoverflow.com/questions/45671803/how-to-use-pytest-to-assert-no-warning-is-raised said that this catch_warnings() call was the pytest >7 way to confirm that no warnings were issued (which is what this code was trying to do). 

The simplefilter("error") turns all warnings into errors, and the catch_warnings() forces it to handle them. The warns(None) stuff was deprecated in pytest7 and now is gone in pytest8.

File: setup.py (hunk 1 above)
Comment (Chris-Sigopt) [reply setup.py:52]: When I ran it I got a whole bunch of errors, but that actually seems to have been another issue- I just tested and it worked fine. Will kill these checks and just unpin pytest in the dev file.

**Review by tjs-intel — APPROVED** [review]

**Pull request merged**